### DISCUSSION TITLE
Wiznet chip SPI speed

### DISCUSSION THREAD

**Discussion by B00mb0x** [micropython/micropython#15294] [ESP32] [answered]
I am trying to connect a [Wiznet W5500 shield](https://docs.arduino.cc/hardware/mkr-eth-shield/) to an ESP32-S3 board using jumper wires for prototyping. I followed the instructions in #11474 and compiled and uploaded a custom firmware. I used the code by @robert-hh from #12820 in order to evaluate if the code is working. Unfortunately lan.isconnected() never returns true. I tried using micropython version 1.22 and 1.23 with similar outcomes.

When looking at the SPI clock with an oscilloscope I can see a 40MHz signal roughly once every second, however the SPI baudrate is only set to 500kHz and when I try using spi.write() for debugging it runs on 500kHz. I even added `#define MICROPY_HW_WIZNET_SPI_BAUDRATE  (500000`) in mpconfigboard.h with no effect.
Is there a way to slow down the baudrate of the network library? 40MHz is too fast for my jumper wire setup.

mpconfigboard.cmake:

```
set(IDF_TARGET esp32s3)

set(SDKCONFIG_DEFAULTS
    boards/sdkconfig.base
    boards/sdkconfig.usb
    boards/sdkconfig.ble
    boards/sdkconfig.spiram_sx
    boards/SIPLOG_ESP32_S3/sdkconfig.board
)

if(MICROPY_BOARD_VARIANT STREQUAL "SPIRAM_OCT")
    set(SDKCONFIG_DEFAULTS
        ${SDKCONFIG_DEFAULTS}
        boards/sdkconfig.240mhz
        boards/sdkconfig.spiram_oct
    )

    list(APPEND MICROPY_DEF_BOARD
        MICROPY_HW_BOARD_NAME="Generic ESP32S3 module with Octal-SPIRAM"
    )
endif()
```

mpconfigboard.h:
```
#ifndef MICROPY_HW_BOARD_NAME
// Can be set by mpconfigboard.cmake.
#define MICROPY_HW_BOARD_NAME               "SIPLOG ESP32-S3 BR500k"
#endif
#define MICROPY_HW_MCU_NAME                 "ESP32S3"

// Enable UART REPL for modules that have an external USB-UART and don't use native USB.
#define MICROPY_HW_ENABLE_UART_REPL         (1)

#define MICROPY_HW_I2C0_SCL                 (9)
#define MICROPY_HW_I2C0_SDA                 (8)

//SPI issues
#define MICROPY_HW_WIZNET_SPI_BAUDRATE  (500000)

```
sdkconfig.board:
```
CONFIG_ESPTOOLPY_FLASHMODE_QIO=y
CONFIG_ESPTOOLPY_FLASHFREQ_80M=y
CONFIG_ESPTOOLPY_AFTER_NORESET=y

CONFIG_ESPTOOLPY_FLASHSIZE_4MB=
CONFIG_ESPTOOLPY_FLASHSIZE_8MB=y
CONFIG_ESPTOOLPY_FLASHSIZE_16MB=
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions-8MiB.csv"

CONFIG_COMPILER_OPTIMIZATION_SIZE=y
CONFIG_ETH_USE_SPI_ETHERNET=y
CONFIG_ETH_SPI_ETHERNET_W5500=y
```

main.py:
```
from machine import Pin, SPI
import network
import socket
import utime

sck = Pin(12)
mosi = Pin(11)
miso = Pin(13)
cs = Pin(14, mode=Pin.OUT, value=1)
rst = Pin(7)

spi = SPI(2,baudrate=500,sck=sck,mosi=mosi,miso=miso)
#spi = SoftSPI(baudrate=500000,sck=sck,mosi=mosi,miso=miso)

lan = network.LAN(phy_type=network.PHY_W5500, phy_addr=1,spi=spi,cs=cs,int=rst, ref_clk_mode = Pin.OUT)
lan.active(True)
while not lan.isconnected():
    print(".", end="")
    utime.sleep(0.25)
```

**Comment by robert-hh** [comment]
Looking at the schematics, the W5500 INT signal is not connected to the board pins. The signal that is connected is RST (Reset), which should be kept high and pulled low for reset. So you have to configure `reset=rst`, and not `int=rst`. refclk_mode and phy_addr may both be not applicable, but do not harm.

**Reply by B00mb0x** [reply]
You're right! I missed that. However if I change the `int=rst` to `reset=rst` I get 
```
File "main.py", line 16, in <module>
OSError: esp_eth_driver_install failed with invalid argument
```
The same error appears, if I completely remove `int=rst`. My solution for now is to keep the code the way it was and just disconnect the line to the rst pin, as it's pulled up internally. 
However the SPI speed is still 40Mhz instead of the lower SPI speed specified. I know the 500 in the code is very low, this was a desperate attempt to get clock speed down. Also lowering the frequency using `machine.freq(40000000)` did not result in the communication getting slower, using `machine.freq(20000000)` the program crashes again due to the same error as shown above.

**Reply by robert-hh** [reply]
The schematics show no INT connection, but the silk screen of the board shows INT at pin D7 of the breakout. So maybe that works. 
In network_lan.c there is a configuration for SPI. It takes values from mpconfigport.h line 303/305 for either 12 or 36 MHz. At the S3 it's 36 MHz. So you could change the setting there MICROPY_PY_NETWORK_LAN_SPI_CLOCK_SPEED_MZ to 1 (=1MHZ).

**Reply by B00mb0x** [reply]
Thanks a lot for the help, MICROPY_PY_NETWORK_LAN_SPI_CLOCK_SPEED_MZ was the clue I needed to get the baud rate down. I had to put `lan.config` and `lan.ifconfig` before `lan.isconnected()`, but now isconnected returns true. However when I try to connect to a socket on my laptop I get an OSError:
```
.........Traceback (most recent call last):
  File "<stdin>", line 31, in <module>
OSError: -201
```
Do you know what -201 means? I couldn't find anything on that.

Right now the lan cable is connected to an access point with DHCP, I tried commenting out `lan.ifconfig` hoping that DHCP would provide the IP address, but then it didn't connect at all.
Here is the code I used:
```
from machine import Pin, SPI
import network
import socket
import utime

sck = Pin(12)
mosi = Pin(11)
miso = Pin(13)
lora_cs = Pin(21, mode=Pin.OUT, value=1) #LoRa chip on same SPI bus
cs = Pin(14, mode=Pin.OUT, value=1)
rst = Pin(7)

spi = SPI(1,sck=sck,mosi=mosi,miso=miso)

lan = network.LAN(phy_type=network.PHY_W5500, phy_addr=1,spi=spi,cs=cs, int=rst)
if not lan.active():
    lan.active(True)
    
lan.config(mac=b'mymacinhex')
lan.ifconfig(('192.168.0.126', '255.255.255.0', '192.168.0.3', '192.168.0.3'))

while not lan.isconnected():
    print(".", end="")
    utime.sleep(0.25)

HOST = "192.168.0.131"  # The server's hostname or IP address
PORT = 65828  # The port used by the server

# Create STREAM TCP socket
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect((HOST, PORT))
s.write("Hello World")
```

**Reply by robert-hh** [reply]
If you set the IP address manually, it looks like the device is connected, but it may not. That is most likely the state you see here. You have a connection if you can "ping" your device from another PC. Some indications are the activity LEDs in the LAN connector. They should blink. If not, then the W5500 is not running or the physical LAN connection is faulty.

**Reply by B00mb0x** [reply]
I decided to put the W5500 Board back on the MKR ZERO and it worked, so the board seems to work properly.
Putting it back to the ESP32 with Micropython, the green activity LED is on constantly and the orange LED at the LAN connector is blinking.
If I comment out the `lan.ifconfig(('192.168.0.126', '255.255.255.0', '192.168.0.3', '192.168.0.3'))`, it never connects, if I leave it in, no messages can be seen on wireshark. I checked the wiring and it seems to be right. Are there any other tests I can perform?

**Reply by robert-hh** [reply]
I tried to replicate your set-up with an ESP32-S3 board and a W5500 breakout, with the same results. But if I set the IP address manually before or after lan.active(), there is a connection. I can "ping" it from my PC, and send out ping requests. I did not try more.
Looking into the code, the call for lan.ifconfig("dhcp") looks fine, and it calls the respective ESP API. Only it does not get an IP configuration. Looks like something is wrong in the esp-idf abyss.

**Reply by B00mb0x** [reply]
Thanks so much for all the effort!
Right now I can see ARP messages sent by the ESP32-S3/W5500 but apparently it doesn't receive or process them, as it keeps asking but can't be pinged. So apparently something works. Could it be an issue with ifconfig? Right now I have a testsetup with a router in between where I use `lan.ifconfig(('192.168.0.78', '255.255.255.0', 'IP of Router', '8.8.8.8'))` and I also test with a direct connection to a windows computer using `lan.ifconfig(('192.168.0.78', '255.255.255.0', 'IP of Windows machine', '8.8.8.8'))`.
![Wireshark](https://github.com/micropython/micropython/assets/20295167/ff8d63f0-6297-404e-8b48-1e45c23c4ed5)
I also noticed, that the code only runs properly after a hard reset. I used soft resets often, but apparently, they don't reset the network stack properly causing some misbehaviours.

**Reply by robert-hh** [reply]
> that the code only runs properly after a hard reset. I

I noticed that as well. Calling lan.active(False) before soft reset does not the soft reset fault. The next thing I could do is connect the w5500 to a plain ESP32. But not today to tomorrow.

**Reply by B00mb0x** [reply]
Did you find some time to test the behaviour on an ESP32?

And do you have any recommandations regarding my setup? If the ARP messages come to my computer something has to work, right? But why doesn't the ESP establish a connection with my computer and keeps sending only the ARP messages?

**Reply by robert-hh** [reply]
Sorry no. I did not test the ESP32 yet. 
But in my test the connection was established if I set the IP address manually. Only DHCP fails.

**Reply by glenn20** [reply]
Thanks @robert-hh and @B00mb0x: following the advice and hints in this discussion and  #12820 I was able to get ethernet working on my [LILYGO T-ETH-LITE](https://www.lilygo.cc/products/t-eth-lite?variant=43120880779445) using the standard `ESP32_GENERIC_S3-SPIRAM_OCT-20240602-v1.23.0.bin` image (no special compiling needed).

**Reply by robert-hh** [reply]
Since @glenn20 mentioned it, I had already connected an esp32 using a w5500. SO looking at my files I found the script, which attached below. The pins used in that script so not matter. An  I tested that again.  What maked a difference is the line:
`nic.config(mac=b'\x02\x00\x00\x12\x34\x56')`

Without that line DHCP did not succeed.~~So you could try adding a similar line to your test file. The specific MAC address does not matter, if it starts with \x02.~~ Do not do that. I tried and it seems I made the W5500 inoperable.

```
import network
import machine
import time
spi = machine.SPI(1, sck=machine.Pin(18), mosi=machine.Pin(21), miso=machine.Pin(19))
nic = network.LAN(phy_type=network.PHY_W5500, spi=spi, phy_addr=1,
                    cs=machine.Pin(5), int=machine.Pin(4))

nic.config(mac=b'\x02\x00\x00\x12\x34\x56')
print("connected:", nic.isconnected())
print('status', nic.status())
network.hostname("new_hostname")
nic.active(1)
print("connected:", nic.isconnected())
print('status', nic.status())

while not nic.isconnected():
    print(".", end="")
    time.sleep(0.25)

print("\nconnected:", nic.isconnected())
print('status', nic.status())
print('mac', nic.config('mac'))
print('ip4 config', nic.ifconfig())
```

**Reply by robert-hh** [reply]
Using `lan.config(mac=b'\x02\x04\x04\x12\x34\x56')` after lan instantiation and before lan.active() worked on my S3 board. The script is now:
```
from machine import Pin, SPI
import network
import socket
import utime

sck = Pin(12)
mosi = Pin(11)
miso = Pin(13)
cs = Pin(14, mode=Pin.OUT, value=1)
rst = Pin(10, Pin.OUT, value=1)
int = Pin(9, Pin.IN)

spi = SPI(2,baudrate=8_000_000,sck=sck,mosi=mosi,miso=miso)
#spi = SoftSPI(baudrate=500000,sck=sck,mosi=mosi,miso=miso)

lan = network.LAN(phy_type=network.PHY_W5500, phy_addr=1,spi=spi,cs=cs,int=int)
lan.config(mac=b'\x02\x04\x04\x12\x34\x56')
lan.active(True)
while not lan.isconnected():
    print(".", end="")
    utime.sleep(0.25)
```

**Reply by glenn20** [reply]
And re. the soft resets, if you put a `nic.active(0)` in your init script, it connects just fine after a soft reset, eg. (adapted from Robert's example code):

```py
import time
import machine
import network

SPI_PINS = dict(sck=10, miso=11, mosi=12)
LAN_PINS = dict(cs=9, int=13, reset=14)
IFCONFIG = ("192.168.10.99", "255.255.255.0", "192.168.10.1", "192.168.10.1")

spi = machine.SPI(1, **SPI_PINS)
lan = network.LAN(phy_type=network.PHY_W5500, spi=spi, phy_addr=1, **LAN_PINS)

# Ensure lan is inactive after a soft reset (will throw OSError if after hard reset)
if machine.reset_cause() == machine.SOFT_RESET:
    lan.active(0)  # Ensure off after soft reset
lan.ifconfig(IFCONFIG)

def connect(lan):
    lan.active(1)
    while not lan.isconnected():
        print(".", end="")
        time.sleep(0.25)
    print()
    print(f"Connected on {lan.ifconfig()[0]}")

connect(lan)
```

**Reply by glenn20** [reply]
> Using `lan.config(mac=b'\x02\x04\x04\x12\x34\x56')` after lan instantiation and before lan.active() worked on my S3 board. 

Thanks for the tip. The esp32s3 (and other esp32 devices) reserve 4 sequential mac addresses for use by the device. The fourth one is reserved for the LAN. See https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32s3/api-reference/system/system.html#mac-address.

`machine.unique_id()` will return the base mac address (without having to startup the wifi interfaces).

So, I now use:

```py
mac = bytearray(machine.unique_id())
mac[-1] += 3  # Add 3 to last octet of base mac to get lan mac
lan.config(mac=mac)
```

**Reply by glenn20** [reply]
Turns out that dhcp servers seem to ignore requests from mac address: b`\x00\x00\x00\x00\x00\x00` ;-)

**Reply by robert-hh** [reply]
Thanks. That's a good hint.

**Reply by glenn20** [reply]
I have submitted PR #15385 to stop `lan.active(True/False)` raising `OSError` if the lan is already in the desired state.

**Comment by B00mb0x** [comment, marked as answer]
I have ordered two new W5500 development boards: the [ETH Wiz click](https://www.mikroe.com/eth-wiz-click) and the [W5500 Lite](https://joy-it.net/en/products/SBC-USR-ES1) board. For some reason, the Arduino shield and the ETH Wiz click still don't work properly. However the W5500 board works using the same code. Thanks for all the help!

**Reply by liudr** [reply]
I know you've found your solution but do you feel like you owe the helpers a clear answer whether the "int=rst" was the source of your error, not the different boards or any steps in  #11474?

**Reply by B00mb0x** [reply]
I cannot say for sure what the error was. All I can say is, that it now works with the W5500 Lite board, but not with the Wiz click board and also not with the Arduino MKR shield. I guess it's not related to the int=rst thing, as the click board also has an interrupt pin, but still doesn't work. My guess is, that is has something to do with the pmode resistors. However I cannot confirm this theory, as I can't find the schematic for the W5500 Lite board.

**Reply by liudr** [reply]
Doing the int=rst is simply wrong! Maybe you happened upon a pin configuration on the "working" board where your defined "rst" variable was indeed containing the pin number connected to W5500's int pin on the "working" board (pin 7?) or just wired Ethernet's int pin to your ESP32 pin 7 by sheer chance. Hopefully later people will read this post carefully instead of glancing at part of this post and just conclude that the click board or Arduino shield doesn't work.

**Reply by B00mb0x** [reply]
I see, sorry I didn't describe my solution detailed enough. I used int=rst because the Arduino board doesn't have the int pin. As discussed in a previous answer, I had to define the int pin, but on the physical wiring I disconnected it, so basically int was connected to nothing.
With the two newer boards I failed to mention, that I connected the int pin of the board to the according GPIO on the ESP32.
